pilot/pkg/serviceregistry/kube/controller/ambient/services.go

```diff
@@ -26,37 +27,145 @@ import (
 	"istio.io/istio/pilot/pkg/model"
 	"istio.io/istio/pilot/pkg/networking/serviceentry"
 	"istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pilot/pkg/serviceregistry/kube/controller/ambient/multicluster"
 	"istio.io/istio/pilot/pkg/util/protoconv"
+	"istio.io/istio/pkg/cluster"
 	"istio.io/istio/pkg/config"
 	"istio.io/istio/pkg/config/host"
 	"istio.io/istio/pkg/config/schema/gvk"
 	"istio.io/istio/pkg/config/schema/kubetypes"
 	"istio.io/istio/pkg/kube/controllers"
 	"istio.io/istio/pkg/kube/krt"
 	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/network"
 	"istio.io/istio/pkg/ptr"
 	"istio.io/istio/pkg/slices"
 	"istio.io/istio/pkg/workloadapi"
 )
 
 func (a *index) ServicesCollection(
+	clusterID cluster.ID,
 	services krt.Collection[*v1.Service],
 	serviceEntries krt.Collection[*networkingclient.ServiceEntry],
 	waypoints krt.Collection[Waypoint],
 	namespaces krt.Collection[*v1.Namespace],
 	opts krt.OptionsBuilder,
 ) krt.Collection[model.ServiceInfo] {
 	ServicesInfo := krt.NewCollection(services, a.serviceServiceBuilder(waypoints, namespaces),
-		opts.WithName("ServicesInfo")...)
+		append(
+			opts.WithName("ServicesInfo"),
+			krt.WithMetadata(krt.Metadata{
```

Comment:
can we make `opts` have a `WithMetadata` func and move this cluster ID piece up a level so it doesn't have to be done in each place (just have it in the shared `opts` we pass into the various XYZCollection funcs

This seems like a step that could easily be forgotten when someone adds something and isnt thinking about multi-cluster.

Reply:
Check out the commit message for 4d72fce218f388917be3ca3d23166400ae5daf39. I tried this approach, but because of the way we pass metadata around, we end up in a worse state because collections get the _wrong_ metadata vs. no metadata. We can catch no metadata in unit tests since I think we panic, but having the wrong metadata leads to tricky bugs